Mathlib/NumberTheory/Rayleigh.lean

```diff
@@ -0,0 +1,187 @@
+`Irrational.beattySeq_symmDiff_beattySeq_pos`, which is Rayleigh's theorem.
+
+## Main definitions
+
+* `beattySeq`: In the Beatty sequence for real number `r`, the `k`th term is `⌊k * r⌋`.
+* `beattySeq'`: In this variant of the Beatty sequence for `r`, the `k`th term is `⌈k * r⌉ - 1`.
+
+## Main statements
+
+Define the following Beatty sets, where `r` denotes a real number:
+
+* `B_r := {⌊k * r⌋ | k ∈ ℤ}`
+* `B'_r := {⌈k * r⌉ - 1 | k ∈ ℤ}`
+* `B⁺_r := {⌊r⌋, ⌊2r⌋, ⌊3r⌋, ...}`
+* `B⁺'_r := {⌈r⌉-1, ⌈2r⌉-1, ⌈3r⌉-1, ...}`
+
+The main statements are:
+
+* `compl_beattySeq`: Let `r` be a real number greater than 1, and `1/r + 1/s = 1`.
+  Then the complement of `B_r` is `B'_s`.
+* `beattySeq_symmDiff_beattySeq'_pos`: Let `r` be a real number greater than 1, and `1/r + 1/s = 1`.
+  Then `B⁺_r` and `B⁺'_s` partition the positive integers.
+* `Irrational.beattySeq_symmDiff_beattySeq_pos`: Let `r` be an irrational number greater than 1, and
+  `1/r + 1/s = 1`. Then `B⁺_r` and `B⁺_s` partition the positive integers.
+
+## References
+
+* [Wikipedia, *Beatty sequence*](https://en.wikipedia.org/wiki/Beatty_sequence)
+
+## Tags
+
+beatty, sequence, rayleigh, irrational, floor, positive
+-/
+
+/-- In the Beatty sequence for real number `r`, the `k`th term is `⌊k * r⌋`. -/
+noncomputable def beattySeq (r : ℝ) : ℤ → ℤ :=
+  fun k ↦ ⌊k * r⌋
+
+/-- In this variant of the Beatty sequence for `r`, the `k`th term is `⌈k * r⌉ - 1`. -/
+noncomputable def beattySeq' (r : ℝ) : ℤ → ℤ :=
+  fun k ↦ ⌈k * r⌉ - 1
+
+namespace Beatty
+
+variable {r s : ℝ} (hrs : r.IsConjugateExponent s) {j k : ℤ}
+
+/-- Let `r > 1` and `1/r + 1/s = 1`. Then `B_r` and `B'_s` are disjoint (i.e. no collision exists).
+-/
+private theorem no_collision : Disjoint {beattySeq r k | k} {beattySeq' s k | k} := by
+  rw [Set.disjoint_left]
+  intro j ⟨k, h₁⟩ ⟨m, h₂⟩
+  rw [beattySeq, Int.floor_eq_iff, ← div_le_iff hrs.pos, ← lt_div_iff hrs.pos] at h₁
+  rw [beattySeq', sub_eq_iff_eq_add, Int.ceil_eq_iff, Int.cast_add, Int.cast_one,
+    add_sub_cancel, ← div_lt_iff hrs.symm.pos, ← le_div_iff hrs.symm.pos] at h₂
+  have h₃ := add_lt_add_of_le_of_lt h₁.1 h₂.1
+  have h₄ := add_lt_add_of_lt_of_le h₁.2 h₂.2
+  simp_rw [div_eq_inv_mul, ← right_distrib, inv_eq_one_div, hrs.inv_add_inv_conj, one_mul] at h₃ h₄
+  rw [← Int.cast_one] at h₄
+  simp_rw [← Int.cast_add, Int.cast_lt, Int.lt_add_one_iff] at h₃ h₄
+  exact h₄.not_lt h₃
+
+/-- Let `r > 1` and `1/r + 1/s = 1`. Suppose there is an integer `j` where `B_r` and `B'_s` both
+jump over `j` (i.e. an anti-collision). Then this leads to a contradiction. -/
+private theorem no_anticollision :
+    ¬∃ j k m : ℤ, k * r < j ∧ j + 1 ≤ (k + 1) * r ∧ m * s ≤ j ∧ j + 1 < (m + 1) * s := by
+  intro ⟨j, k, m, h₁₁, h₁₂, h₂₁, h₂₂⟩
+  replace h₁₁ := (lt_div_iff hrs.pos).2 h₁₁
+  replace h₁₂ := (div_le_iff hrs.pos).2 h₁₂
+  replace h₂₁ := (le_div_iff hrs.symm.pos).2 h₂₁
+  replace h₂₂ := (div_lt_iff hrs.symm.pos).2 h₂₂
```

Comment:
```suggestion
  simp only, [←div_le_iff, ←le_div_iff, ←div_lt_iff, ←lt_div_iff, hrs.pos, hrs.symm.pos] at h₁₁ h₁₂ h₂₁ h₂₂
```

Reply:
Done.
(I managed to golf this away in "More golf v2")